File: Data Structures & Algorithms/range-sum-of-bst/submission-3.py
"""
Understand:
Provided a tree and [low, high], return the sum of nodes that are within low/high
Example1, (3,8), 5+3+8+4+7=27
Example2, (2,4), 4+3+2=9
All nodes' values are unique

Match:
BFS, Intervals

Plan:
Do a BFS from root, add the value of node if it's within low/high values inclusively
"""
"""
A Binary Search Tree (BST) is a type of binary tree data structure in which each node contains a unique key and satisfies a specific ordering property:

All nodes in the left subtree of a node contain values strictly less than the node’s value.
All nodes in the right subtree of a node contain values strictly greater than the node’s value.
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        if root is None:
            return 0

        result = 0
        s = []
        s.append(root)

        while s:
            node = s.pop()

            if node is None:
                continue
            if low <= node.val <= high:
                result += node.val
            if node.val > low:
                s.append(node.left)
            if node.val < high:
                s.append(node.right)
            
        return result

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        auto dfs = [&](this auto&& self, TreeNode* node) -> TreeNode* {
            // If current node's value is greater than p's and q's value, go left
            if (node->val > p->val && node->val > q->val) {
                return self(node->left);
            }

            // If current node's value is less than p's and q's value, go right
            if (node->val < p->val && node->val < q->val) {
                return self(node->right);
            }

            // Return the lowest common ancestor
            return node;
        };

        // Return empty if any of root, p, and q is empty
        if (root == nullptr || p == nullptr || q == nullptr) {
            return nullptr;
        }
        return dfs(root);
    }
};
